File: lib/python-beta/src/bailo/helper/access_request.py
from __future__ import annotations

from typing import Any

from bailo.core.client import Client


class AccessRequest:
    """Represent a review within Bailo.

    A review can either be access to a model or to a specific release.

    :param client: A client object that is used to make requests to bailo
    :param name: The name of the access request
    :param model_id: The unique model id of the model that the access request is being made with
    :param schema_id: An ID for the schema within Bailo
    :param metadata: A metadata object
    :param access_request_id: The Unique ID for this access request
    :param deleted: Whether the access request has been deleted
    """

    def __init__(
        self,
        client: Client,
        model_id: str,
        schema_id: str,
        metadata: Any,
        access_request_id: str,
        created_by: str,
        deleted: bool = False,
    ) -> None:
        self.client = client
        self.model_id = model_id
        self.schema_id = schema_id
        self.metadata = metadata
        self.access_request_id = access_request_id
        self.created_by = created_by
        self.deleted = deleted

    @classmethod
    def from_id(cls, client: Client, model_id: str, access_request_id: str) -> AccessRequest:
        """Return an existing review from Bailo given it's unique ID.

        :param client: A client object used to interact with Bailo
        :param model_id: A unique model ID within Bailo
        :param access_request_id: A unique ID for an access request
        """
        json_access_request = client.get_access_request(model_id, access_request_id)["accessRequest"]

        deleted = json_access_request["deleted"]
        created_by = json_access_request["createdBy"]
        metadata = json_access_request["metadata"]

        schema_id = json_access_request["schemaId"]

        return cls(
            client,
            model_id,
            schema_id,
            metadata,
            access_request_id,
            created_by,
            deleted,
        )

    @classmethod
    def create(cls, client: Client, model_id: str, schema_id: str, metadata: Any) -> AccessRequest:
        """Make an access request for the model.

        Posts an access request to Bailo to be reviewed

        :param client: A client object used to interact with Bailo
        :param name: The name of the access request
        :param model_id: A unique model ID within Bailo
        :param schema_id: A unique schema ID
        :return: JSON response object
        """
        access_request_json = client.post_access_request(model_id, metadata, schema_id)["accessRequest"]

        deleted = access_request_json["deleted"]
        access_request_id = access_request_json["id"]
        metadata = access_request_json["metadata"]
        created_by = access_request_json["createdBy"]

        return cls(
            client,
            model_id,
            schema_id,
            metadata,
            access_request_id,
            created_by,
            deleted,
        )

    def delete(self) -> bool:
        """Delete the access request on Bailo.

        :return: A message confirming the removal of the access request.
        """
        self.client.delete_access_request(self.model_id, self.access_request_id)
        return True

    def update(self):
        """Update the current state of the access request to Bailo."""
        self.client.patch_access_request(self.model_id, self.access_request_id, metadata=self.metadata)

    def __str__(self) -> str:
        return f"Access Request: {self.metadata['overview']['name']} - {self.model_id}"

    def __repr__(self):
        return f"{self.__class__.__name__}({self.model_id},{self.schema_id})"

    def __eq__(self, other):
        if not isinstance(other, self.__class__):
            return NotImplemented
        return self.access_request_id == other.access_request_id

File: backend/src/services/accessRequest.ts
import { Validator } from 'jsonschema'
import { ClientSession, PipelineStage, Types } from 'mongoose'

import { Roles } from '../connectors/authentication/constants.js'
import authentication from '../connectors/authentication/index.js'
import { AccessRequestAction } from '../connectors/authorisation/actions.js'
import authorisation from '../connectors/authorisation/index.js'
import AccessRequestModel, { AccessRequestDoc, AccessRequestInterface } from '../models/AccessRequest.js'
import { ModelDoc } from '../models/Model.js'
import ResponseModel, { ResponseKind } from '../models/Response.js'
import ReviewModel from '../models/Review.js'
import { UserInterface } from '../models/User.js'
import { WebhookEvent } from '../models/Webhook.js'
import { AccessRequestUserPermissions } from '../types/types.js'
import { isValidatorResultError } from '../types/ValidatorResultError.js'
import { toEntity } from '../utils/entity.js'
import { BadReq, Forbidden, InternalError, NotFound } from '../utils/error.js'
import { convertStringToId } from '../utils/id.js'
import { authResponseToUserPermission } from '../utils/permissions.js'
import log from './log.js'
import { getModelById } from './model.js'
import { removeResponsesByParentIds } from './response.js'
import { createAccessRequestReviews, removeAccessRequestReviews } from './review.js'
import { getSchemaById } from './schema.js'
import { sendWebhooks } from './webhook.js'

export type CreateAccessRequestParams = Pick<AccessRequestInterface, 'metadata' | 'schemaId'>
export async function createAccessRequest(
  user: UserInterface,
  modelId: string,
  accessRequestInfo: CreateAccessRequestParams,
) {
  // Check the model exists and the user can view it before creating an access request
  const model = await getModelById(user, modelId)

  // Ensure that the AR meets the schema
  const schema = await getSchemaById(accessRequestInfo.schemaId)
  if (schema.hidden) {
    throw BadReq('Cannot create new Access Request using a hidden schema.', { schemaId: accessRequestInfo.schemaId })
  }
  try {
    new Validator().validate(accessRequestInfo.metadata, schema.jsonSchema, { throwAll: true, required: true })
  } catch (error) {
    if (isValidatorResultError(error)) {
      throw BadReq('Access Request Metadata could not be validated against the schema.', {
        schemaId: accessRequestInfo.schemaId,
        validationErrors: error.errors,
      })
    }
    throw error
  }

  const accessRequestId = convertStringToId(accessRequestInfo.metadata.overview.name)
  const accessRequest = new AccessRequestModel({
    id: accessRequestId,
    createdBy: user.dn,
    modelId,
    comments: [],
    ...accessRequestInfo,
  })

  const auth = await authorisation.accessRequest(user, model, accessRequest, AccessRequestAction.Create)
  if (!auth.success) {
    throw Forbidden(auth.info, { userDn: user.dn, accessRequestId })
  }

  await accessRequest.save()

  try {
    await createAccessRequestReviews(model, accessRequest)
  } catch (error) {
    // Transactions here would solve this issue.
    log.warn(error, 'Error when creating Release Review Requests. Approval cannot be given to this Access Request')
  }

  sendWebhooks(
    accessRequest.modelId,
    WebhookEvent.CreateAccessRequest,
    `Access Request ${accessRequest.id} has been created for model ${accessRequest.modelId}`,
    { accessRequest },
  )

  return accessRequest
}

export async function removeAccessRequests(
  user: UserInterface,
  accessRequestIds: string[],
  session?: ClientSession | undefined,
) {
  // Model cache
  const models: Record<string, ModelDoc> = {}

  for (const accessRequestId of accessRequestIds) {
    const accessRequest = await getAccessRequestById(user, accessRequestId)
    let model: ModelDoc
    if (accessRequest.modelId in models) {
      model = models[accessRequest.modelId]
    } else {
      model = await getModelById(user, accessRequest.modelId)
      models[accessRequest.modelId] = model
    }

    const auth = await authorisation.accessRequest(user, model, accessRequest, AccessRequestAction.Delete)
    if (!auth.success) {
      throw Forbidden(auth.info, { userDn: user.dn, accessRequestId })
    }

    const reviewsForAccessRequest = await ReviewModel.find({ accessRequestId })

    await accessRequest.delete(session)
    await removeAccessRequestReviews(accessRequestId, session)
    await removeResponsesByParentIds(
      [...reviewsForAccessRequest.map((review) => review['_id']), accessRequest['_id']] as string[],
      session,
    )
  }

  return { accessRequestIds }
}

export async function removeAccessRequest(
  user: UserInterface,
  accessRequestId: string,
  session?: ClientSession | undefined,
) {
  await removeAccessRequests(user, [accessRequestId], session)

  return { accessRequestId }
}

export async function getAccessRequestsByModel(user: UserInterface, modelId: string) {
  const model = await getModelById(user, modelId)
  const accessRequests = await AccessRequestModel.find({ modelId })

  const auths = await authorisation.accessRequests(user, model, accessRequests, AccessRequestAction.View)
  return accessRequests.filter((_, i) => auths[i].success)
}

export async function getAccessRequestById(user: UserInterface, accessRequestId: string) {
  const accessRequest = await AccessRequestModel.findOne({ id: accessRequestId })
  if (!accessRequest) {
    throw NotFound('The requested access request was not found.', { accessRequestId })
  }

  const model = await getModelById(user, accessRequest.modelId)

  const auth = await authorisation.accessRequest(user, model, accessRequest, AccessRequestAction.View)
  if (!auth.success) {
    throw Forbidden(auth.info, { userDn: user.dn, accessRequestId })
  }

  return accessRequest
}

export async function findAccessRequests(
  user: UserInterface,
  modelIds: Array<string>,
  schemaId: string,
  mine: boolean,
  adminAccess?: boolean,
): Promise<Array<AccessRequestDoc>> {
  if (adminAccess) {
    if (!(await authentication.hasRole(user, Roles.Admin))) {
      throw Forbidden('You do not have the required role.', {
        userDn: user.dn,
        requiredRole: Roles.Admin,
      })
    }
  }

  const query: any = {}

  if (modelIds.length) {
    query.modelId = { $in: modelIds }
  }

  if (schemaId) {
    query.schemaId = { $eq: schemaId }
  }

  if (mine) {
    query['metadata.overview.entities'] = {
      $in: await authentication.getEntities(user),
    }
  }

  const stages: PipelineStage[] = [{ $match: query }]

  //Auth already checked, so just need to check if they require admin access
  if (adminAccess) {
    const results = await AccessRequestModel.aggregate(stages)
    return results
  }

  stages.push(
    { $group: { _id: '$modelId', accessRequests: { $push: '$$ROOT' } } },
    {
      $lookup: {
        from: 'v2_models',
        localField: '_id',
        foreignField: 'id',
        as: 'model',
      },
    },
    {
      $unwind: '$model',
    },
  )

  const results = await AccessRequestModel.aggregate(stages)

  const accessRequests: AccessRequestDoc[] = []
  for (const result of results) {
    const auth = await authorisation.accessRequests(user, result.model, result.accessRequests, AccessRequestAction.View)

    const authorisedIds = new Set(auth.filter((result) => result.success).map((result) => result.id))
    const filteredAccessRequests = result.accessRequests.filter((accessRequest) => authorisedIds.has(accessRequest.id))
    accessRequests.push(...filteredAccessRequests)
  }
  return accessRequests
}

export type UpdateAccessRequestParams = Pick<AccessRequestInterface, 'metadata'>
export async function updateAccessRequest(
  user: UserInterface,
  accessRequestId: string,
  diff: Partial<UpdateAccessRequestParams>,
) {
  const accessRequest = await getAccessRequestById(user, accessRequestId)
  const model = await getModelById(user, accessRequest.modelId)

  const auth = await authorisation.accessRequest(user, model, accessRequest, AccessRequestAction.Update)
  if (!auth.success) {
    throw Forbidden(auth.info, { userDn: user.dn, accessRequestId })
  }

  // Ensure that the AR meets the schema
  const schema = await getSchemaById(accessRequest.schemaId)
  try {
    new Validator().validate(accessRequest.metadata, schema.jsonSchema, { throwAll: true, required: true })
  } catch (error) {
    if (isValidatorResultError(error)) {
      throw BadReq('Access Request Metadata could not be validated against the schema.', {
        schemaId: accessRequest.schemaId,
        validationErrors: error.errors,
      })
    }
    throw error
  }

  if (diff.metadata) {
    accessRequest.metadata = diff.metadata
    accessRequest.markModified('metadata')
  }

  await accessRequest.save()

  return accessRequest
}

export async function newAccessRequestComment(user: UserInterface, accessRequestId: string, message: string) {
  const accessRequest = await AccessRequestModel.findOne({ id: accessRequestId })

  if (!accessRequest) {
    throw NotFound(`The requested access request was not found.`, { accessRequestId })
  }

  const commentResponse = new ResponseModel({
    entity: toEntity('user', user.dn),
    kind: ResponseKind.Comment,
    comment: message,
    parentId: accessRequest._id as Types.ObjectId,
    createdAt: new Date().toISOString(),
  })

  const newComment = await commentResponse.save()

  if (!newComment) {
    throw InternalError('There was a problem saving a new comment for this access request')
  }

  return commentResponse
}

export async function getModelAccessRequestsForUser(user: UserInterface, modelId: string) {
  const accessRequests = await AccessRequestModel.find({
    modelId,
    'metadata.overview.entities': { $in: await authentication.getEntities(user) },
  })

  return accessRequests
}

export async function getCurrentUserPermissionsByAccessRequest(
  user: UserInterface,
  accessRequestId: string,
): Promise<AccessRequestUserPermissions> {
  const accessRequest = await getAccessRequestById(user, accessRequestId)
  const model = await getModelById(user, accessRequest.modelId)

  const editAccessRequestAuth = await authorisation.accessRequest(
    user,
    model,
    accessRequest,
    AccessRequestAction.Update,
  )
  const deleteAccessRequestAuth = await authorisation.accessRequest(
    user,
    model,
    accessRequest,
    AccessRequestAction.Delete,
  )

  return {
    editAccessRequest: authResponseToUserPermission(editAccessRequestAuth),
    deleteAccessRequest: authResponseToUserPermission(deleteAccessRequestAuth),
  }
}
